// root-module/lib/index.js
'use strict';

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _fs = require('fs');

var _fs2 = _interopRequireDefault(_fs);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function findRootModule(mod) {
  var m = mod || module;
  return m.parent ? findRootModule(m.parent) : m;
}

function findFile(curDir, filename) {
  var fullpath = _path2.default.resolve(curDir, filename);
  try {
    if (_fs2.default.statSync(fullpath).isFile()) {
      return fullpath;
    }
  } catch (e) {
    // not a file
  }

  var par = _path2.default.resolve(curDir, '..');
  if (par === curDir) {
    return null;
  }

  return findFile(par, filename);
}

function rootModule(mod) {
  var m = findRootModule(mod);
  var folder = _path2.default.dirname(m.filename);
  return {
    module: m,
    folder: folder,
    packageFile: findFile(folder, 'package.json'),
    findFile: findFile.bind(null, folder)
  };
}

module.exports = rootModule;
